src/yarp.c

```diff
@@ -11218,42 +11235,112 @@ parse_expression_prefix(yp_parser_t *parser, yp_binding_power_t binding_power) {
             yp_heredoc_indent_t indent = parser->lex_modes.current->as.heredoc.indent;
 
             yp_node_t *node;
-            if (quote == YP_HEREDOC_QUOTE_BACKTICK) {
-                node = (yp_node_t *) yp_interpolated_xstring_node_create(parser, &parser->current, &parser->current);
-            } else {
-                node = (yp_node_t *) yp_interpolated_string_node_create(parser, &parser->current, NULL, &parser->current);
-            }
 
             parser_lex(parser);
-            yp_node_t *part;
+
+            if (parser->current.type == YP_TOKEN_HEREDOC_END) {
+                if (quote == YP_HEREDOC_QUOTE_BACKTICK) {
+                    yp_x_string_node_t *x_str_node = yp_xstring_node_create(parser, &parser->previous, &parser->previous, &parser->current);
+                    x_str_node->content_loc = (yp_location_t) {
+                        .start = parser->previous.end,
+                        .end = parser->current.start
+                    };
+                    x_str_node->base.location = (yp_location_t) {
+                        .start = parser->previous.start,
+                        .end = parser->previous.end
+                    };
+                    node = (yp_node_t *) x_str_node;
+                } else {
+                    yp_string_node_t *str_node = yp_string_node_create(parser, &parser->previous, &parser->previous, &parser->current);
+                    str_node->content_loc = (yp_location_t) {
+                        .start = parser->previous.end,
+                        .end = parser->current.start
+                    };
+                    str_node->base.location = (yp_location_t) {
+                        .start = parser->previous.start,
+                        .end = parser->previous.end
+                    };
+                    node = (yp_node_t *) str_node;
+                }
+                lex_state_set(parser, YP_LEX_STATE_END);
+                expect(parser, YP_TOKEN_HEREDOC_END, YP_ERR_HEREDOC_TERM);
+                return node;
+            }
+
+            yp_token_t opening_token = parser->previous;
+
+            yp_node_t *part = NULL;
+            yp_node_list_t parts = YP_EMPTY_NODE_LIST;
```

Comment:
Here you're creating a list, then appending all of the parts, then checking the size of the list. I think this might be easier to read if you instead parsed the first part, checked its contents and the next token, and then only created a list if it was necessary.

For example:

```c
yp_node_t *part = parse_string_part(parser);
if (part == NULL) {
    // We couldn't parse anything, so just return a missing node.
    return (yp_node_t *) yp_missing_node_create(parser, parser->previous.start, parser->previous.end);
}

if (YP_NODE_TYPE(part, YP_STRING_NODE) && match_any_type_p(parser, 2, YP_TOKEN_HEREDOC_END, YP_TOKEN_EOF)) {
    // Here we only have a single string, so we can return just the one string part.
}

// Otherwise, we need to parse all of the parts of the string.
yp_node_list_t parts = YP_EMPTY_NODE_LIST;
yp_node_list_append(&parts, part);

while (!match_any_type_p(parser, 2, YP_TOKEN_HEREDOC_END, YP_TOKEN_EOF)) {
    if ((part = parse_string_part(parser)) == NULL) continue;
    yp_node_list_append(&parts, part);
}

// Now we have all the parts of the string, create the correct node and return it.
```